lasertag_ppo.py

```diff
@@ -0,0 +1,352 @@
+
+        return pz_obs
+
+    def step(self, action: dict[AgentID, ActionType], device: str) -> tuple[
+        dict[AgentID, ObsType],
+        dict[AgentID, float],
+        dict[AgentID, bool],
+        dict[AgentID, bool],
+        dict[AgentID, dict],
+    ]:
+        """
+        Takes inputs in the PettingZoo (PZ) Parallel API format, performs a step and
+        returns outputs in PZ format.
+        """
+        action = batchify(action, device)
+        obs, rew, done, info = self.env.step(action)
+        obs = obs["image"]
+        trunc = 0  # there is no `truncated` flag in this environment
+        self.task_completion = self._task_completion(obs, rew, done, trunc, info)
+        # convert outputs back to PZ format
+        obs, rew = tuple(map(self._np_array_to_pz_dict, [obs, rew]))
+        done, trunc, info = tuple(
+            map(self._singleton_to_pz_dict, [done, trunc, self.task_completion])
+        )
+
+        return self.observation(obs), rew, done, trunc, info
+
+
+class SelfPlay(Curriculum):
+    def __init__(self, agent, device: str, store_agents_on_cpu: bool = False):
+        self.store_agents_on_cpu = store_agents_on_cpu
+        self.storage_device = "cpu" if self.store_agents_on_cpu else device
+        self.agent = deepcopy(agent).to(self.storage_device)
+
+    def update_agent(self, agent):
+        self.agent = deepcopy(agent).to(self.storage_device)
+
+    def get_opponent(self):
+        # Always return the most recent agent
+        return self.agent
+
+    def sample(self, k=1):
+        return 0
+
+
+class Agent(nn.Module):
+    def __init__(self, num_actions):
+        super().__init__()
+
+        self.network = nn.Sequential(
+            self._layer_init(nn.Linear(3 * 5 * 5, 512)),
+            nn.ReLU(),
+        )
+        self.actor = self._layer_init(nn.Linear(512, num_actions), std=0.01)
+        self.critic = self._layer_init(nn.Linear(512, 1))
+
+    def _layer_init(self, layer, std=np.sqrt(2), bias_const=0.0):
+        torch.nn.init.orthogonal_(layer.weight, std)
+        torch.nn.init.constant_(layer.bias, bias_const)
+        return layer
+
+    def get_value(self, x, flatten_start_dim=1):
+        x = torch.flatten(x, start_dim=flatten_start_dim)
+        return self.critic(self.network(x / 255.0))
+
+    def get_action_and_value(self, x, action=None, flatten_start_dim=1):
+        x = torch.flatten(x, start_dim=flatten_start_dim)
+        hidden = self.network(x / 255.0)
+        logits = self.actor(hidden)
+        probs = Categorical(logits=logits)
+        if action is None:
+            action = probs.sample()
+        return action, probs.log_prob(action), probs.entropy(), self.critic(hidden)
+
+
+def batchify_obs(obs, device):
+    """Converts PZ style observations to batch of torch arrays."""
+    # convert to list of np arrays
+    obs = np.stack([obs[a] for a in obs], axis=0)
+    # convert to torch
+    obs = torch.tensor(obs).to(device)
+
+    return obs
```

Comment:
Is there any reason not to just use `batchify`? They do the exact same thing right?

Reply:
That's right, the original PettingZoo code had an added `obs = obs.transpose(0, -1, 1, 2)` in `batchify_obs` which I removed. So there's no need for this function anymore